server/src/main/java/org/elasticsearch/search/diversification/ResultDiversificationRetrieverBuilder.java

```diff
@@ -0,0 +1,313 @@
+    private final float[] queryVector;
+    private final Float lambda;
+    private ResultDiversificationContext diversificationContext = null;
+
+    ResultDiversificationRetrieverBuilder(
+        RetrieverSource innerRetriever,
+        String diversificationType,
+        String diversificationField,
+        int rankWindowSize,
+        @Nullable float[] queryVector,
+        @Nullable Float lambda
+    ) {
+        super(List.of(innerRetriever), rankWindowSize);
+        this.diversificationType = diversificationType;
+        this.diversificationField = diversificationField;
+        this.queryVector = queryVector;
+        this.lambda = lambda;
+    }
+
+    @Override
+    protected ResultDiversificationRetrieverBuilder clone(
+        List<RetrieverSource> newChildRetrievers,
+        List<QueryBuilder> newPreFilterQueryBuilders
+    ) {
+        assert newChildRetrievers.size() == 1 : "ResultDiversificationRetrieverBuilder must have a single child retriever";
+        return new ResultDiversificationRetrieverBuilder(
+            newChildRetrievers.getFirst(),
+            diversificationType,
+            diversificationField,
+            rankWindowSize,
+            queryVector,
+            lambda
+        );
+    }
+
+    @Override
+    public ActionRequestValidationException validate(
+        SearchSourceBuilder source,
+        ActionRequestValidationException validationException,
+        boolean isScroll,
+        boolean allowPartialSearchResults
+    ) {
+        // ensure the type is one we know of - at the moment, only "mmr" is valid
+        if (diversificationType.equals(DIVERSIFICATION_TYPE_MMR) == false) {
+            validationException = addValidationError(
+                String.format(Locale.ROOT, "[%s] diversification type must be set to [%s]", getName(), DIVERSIFICATION_TYPE_MMR),
+                validationException
+            );
+        }
+
+        // if MMR, ensure we have a lambda between 0.0 and 1.0
+        if (diversificationType.equals(DIVERSIFICATION_TYPE_MMR) && (lambda == null || lambda < 0.0 || lambda > 1.0)) {
+            validationException = addValidationError(
+                String.format(
+                    Locale.ROOT,
+                    "[%s] MMR result diversification must have a [%s] between 0.0 and 1.0",
+                    getName(),
+                    LAMBDA_FIELD.getPreferredName()
+                ),
+                validationException
+            );
+        }
+
+        return validationException;
+    }
+
+    @Override
+    protected RetrieverBuilder doRewrite(QueryRewriteContext ctx) {
+        IndexVersion indexVersion = ctx.getIndexSettings().getIndexVersionCreated();
+        Mapper mapper = ctx.getMappingLookup().getMapper(diversificationField);
+        if (mapper instanceof DenseVectorFieldMapper == false) {
+            throw new IllegalArgumentException("[" + diversificationField + "] is not a dense vector field");
```

Comment:
Nitpick - Error message should be `supported field types: [dense_vector]` - that leaves the door open for future field types (like semantic_text!) in the future.